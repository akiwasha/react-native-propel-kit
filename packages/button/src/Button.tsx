// @see https://github.com/facebook/react-native/blob/master/Libraries/Components/Button.js
// @see https://github.com/react-native-training/react-native-elements/blob/next/src/buttons/Button.js

import React, {ElementType, FunctionComponent, useCallback, useMemo} from 'react';
import {
  ButtonProps as BaseButtonProps,
  GestureResponderEvent,
  Platform,
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableNativeFeedback,
  TouchableNativeFeedbackProps,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
  ViewStyle
} from 'react-native';
import {pickTextStyles} from './utils';

export const defaultStyles = {
  view: Platform.select<ViewStyle>({
    ios: {justifyContent: 'center', flexDirection: 'row', height: 50},
    android: {justifyContent: 'center', flexDirection: 'row', height: 50}
  }),
  disabledView: Platform.select<ViewStyle>({
    ios: {},
    android: {}
  }),
  text: Platform.select<TextStyle>({
    ios: {textAlign: 'center', margin: 8, color: '#007aff', fontSize: 18},
    android: {paddingHorizontal: 3, paddingBottom: 2, color: '#212121'}
  }),
  disabledText: Platform.select<TextStyle>({
    ios: {color: '#cdcdcd'},
    android: {color: '#cdcdcd'}
  })
  // loading: Platform.select<ViewStyle>({
  //   ios: {
  //     position: 'absolute',
  //     backgroundColor: 'black',
  //     opacity: 0.5,
  //     height: 40,
  //     width: 40,
  //     marginRight: 8,
  //     alignSelf: 'center',
  //     borderRadius: 12
  //   },
  //   android: {marginRight: 4}
  // })
};

export type ButtonProps = BaseButtonProps & {
  // loading?: boolean;
  // loadingStyle?: StyleProp<ViewStyle>;
  style?: StyleProp<TextStyle>;
  viewStyle?: StyleProp<ViewStyle>;
  disabledViewStyle?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  disabledTextStyle?: StyleProp<TextStyle>;
  TouchableComponent?: ElementType<TouchableNativeFeedbackProps | TouchableOpacityProps>;
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  defaultStyles?: typeof defaultStyles;
};

export const defaultProps = {
  TouchableComponent: Platform.select<ElementType<TouchableNativeFeedbackProps | TouchableOpacityProps>>({
    android: TouchableNativeFeedback,
    default: TouchableOpacity
  }),
  defaultStyles: typeof defaultStyles
};

export const Button: FunctionComponent<ButtonProps> = ({
  title,
  disabled,
  // loading: propLoading,
  onPress,
  color,
  viewStyle: propViewStyle,
  // loadingStyle: propLoadingStyle,
  disabledViewStyle,
  textStyle: propTextStyle,
  disabledTextStyle,
  defaultStyles: propDefaultStyles = defaultStyles,
  style: propStyle,
  TouchableComponent = defaultProps.TouchableComponent,
  ...otherTouchableProps
}) => {
  // @NOTE we use flatten to properly split text/view related styles
  const flattenStyle = useMemo<TextStyle>(() => (propStyle ? StyleSheet.flatten(propStyle) : {}), [propStyle]);

  const handlePress = useCallback(
    (ev: GestureResponderEvent) => {
      if (disabled) {
        return;
      }
      if (onPress) {
        onPress(ev);
      }
    },
    [disabled, onPress]
  );

  const viewStyle = useMemo<ViewStyle>(
    () =>
      StyleSheet.flatten([
        propDefaultStyles.view,
        propViewStyle,
        flattenStyle,
        disabled ? [propDefaultStyles.disabledView, disabledViewStyle] : null
      ]),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [propViewStyle, flattenStyle, disabled, disabledViewStyle]
  );

  const textStyle = useMemo<TextStyle>(
    () =>
      StyleSheet.flatten([
        propDefaultStyles.text,
        propTextStyle,
        pickTextStyles(flattenStyle),
        color ? {color} : null,
        disabled ? [propDefaultStyles.disabledText, disabledTextStyle] : null
      ]),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [propTextStyle, flattenStyle, disabled, color, disabledTextStyle]
  );

  return (
    <TouchableComponent
      style={{flex: viewStyle.flex ? 1 : 0, flexDirection: 'row'}}
      disabled={disabled}
      onPress={handlePress}
      {...otherTouchableProps}
    >
      <View style={viewStyle}>
        {/* {propLoading ? (
          <ActivityIndicator
            style={[propDefaultStyles.loading, propLoadingStyle]}
            color={textStyle && textStyle.color ? textStyle.color : propDefaultStyles.text.color}
          />
        ) : null} */}
        <Text style={textStyle}>{title}</Text>
      </View>
    </TouchableComponent>
  );
};
