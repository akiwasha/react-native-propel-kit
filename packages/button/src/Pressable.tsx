// @see https://github.com/facebook/react-native/blob/master/Libraries/Components/Button.js
// @see https://github.com/react-native-training/react-native-elements/blob/next/src/buttons/Button.js

import React, {ElementType, FunctionComponent, ReactNode, useCallback, useMemo} from 'react';
import {
  ButtonProps,
  GestureResponderEvent,
  Platform,
  StyleProp,
  StyleSheet,
  Text,
  TextProps,
  TextStyle,
  TouchableNativeFeedback,
  TouchableNativeFeedbackProps,
  TouchableOpacity,
  TouchableOpacityProps,
  ViewStyle
} from 'react-native';
import {pickTextStyles} from './utils';

export type PressableProps = ButtonProps &
  Pick<TextProps, 'ellipsizeMode' | 'numberOfLines'> & {
    children?: ReactNode;
    style?: StyleProp<TextStyle>;
    viewStyle?: StyleProp<ViewStyle>;
    disabledViewStyle?: StyleProp<ViewStyle>;
    textStyle?: StyleProp<TextStyle>;
    disabledTextStyle?: StyleProp<TextStyle>;
    TouchableComponent?: ElementType<TouchableNativeFeedbackProps | TouchableOpacityProps>;
    activeOpacity?: TouchableOpacityProps['activeOpacity'];
    defaultStyles?: typeof defaultStyles;
  };

export const defaultProps = {
  TouchableComponent: Platform.select<ElementType<TouchableNativeFeedbackProps | TouchableOpacityProps>>({
    android: TouchableNativeFeedback,
    default: TouchableOpacity
  })
};

export const Pressable: FunctionComponent<PressableProps> = ({
  children,
  title,
  disabled,
  onPress,
  color,
  viewStyle: propViewStyle,
  disabledViewStyle,
  textStyle: propTextStyle,
  ellipsizeMode,
  numberOfLines,
  disabledTextStyle,
  defaultStyles: propDefaultStyles = defaultStyles,
  style: propStyle,
  TouchableComponent = defaultProps.TouchableComponent,
  ...otherTouchableProps
}) => {
  // @NOTE we use flatten to properly split text/view related styles
  const flattenStyle = useMemo<TextStyle>(() => (propStyle ? StyleSheet.flatten(propStyle) : {}), [propStyle]);

  const handlePress = useCallback(
    (ev: GestureResponderEvent) => {
      if (disabled) {
        return;
      }
      if (onPress) {
        onPress(ev);
      }
    },
    [disabled, onPress]
  );

  const viewStyle = useMemo<ViewStyle>(
    () =>
      StyleSheet.flatten([propDefaultStyles.view, propViewStyle, flattenStyle, disabled ? disabledViewStyle : null]),
    [propDefaultStyles.view, propViewStyle, flattenStyle, disabled, disabledViewStyle]
  );

  const textStyle = useMemo<TextStyle>(
    () =>
      StyleSheet.flatten([
        propDefaultStyles.text,
        propTextStyle,
        pickTextStyles(flattenStyle),
        color ? {color} : null,
        disabled ? disabledTextStyle : null
      ]),
    [propDefaultStyles.text, propTextStyle, flattenStyle, disabled, color, disabledTextStyle]
  );

  return (
    <TouchableComponent style={viewStyle} disabled={disabled} onPress={handlePress} {...otherTouchableProps}>
      {children}
      <Text ellipsizeMode={ellipsizeMode} numberOfLines={numberOfLines} style={textStyle}>
        {title}
      </Text>
    </TouchableComponent>
  );
};

export const defaultStyles = {
  view: Platform.select<ViewStyle>({
    ios: {justifyContent: 'center', flexDirection: 'row', height: 50},
    android: {justifyContent: 'center', flexDirection: 'row', height: 50}
  }),
  disabledView: Platform.select<ViewStyle>({
    ios: {},
    android: {}
  }),
  text: Platform.select<TextStyle>({
    ios: {alignSelf: 'center', fontSize: 18, color: '#007aff'},
    android: {paddingHorizontal: 3, paddingBottom: 2, color: '#212121'}
  }),
  disabledText: Platform.select<TextStyle>({
    ios: {color: '#cdcdcd'},
    android: {color: '#cdcdcd'}
  })
};
