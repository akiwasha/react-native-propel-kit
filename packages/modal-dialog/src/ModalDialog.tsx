import {BackdropContext, BackdropContextValue} from '@mgcrea/react-native-backdrop-provider';
import React, {
  forwardRef,
  ReactNode,
  useCallback,
  useContext,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState
} from 'react';
import {
  Dimensions,
  Keyboard,
  Modal,
  ModalProps,
  Platform,
  SafeAreaView,
  StyleProp,
  Text,
  TextStyle,
  TouchableWithoutFeedback,
  View,
  ViewStyle
} from 'react-native';
import {ModalDialogButton} from './components';

export type ModalDialogProps = ModalProps & {
  defaultStyles?: typeof defaultStyles;
  backgroundColor?: string;
  cancelTitle?: string;
  cancelStyle?: StyleProp<TextStyle>;
  children?: ReactNode;
  confirmTitle?: string;
  confirmStyle?: StyleProp<TextStyle>;
  modalStyle?: StyleProp<ViewStyle>;
  containerStyle?: StyleProp<ViewStyle>;
  headerStyle?: StyleProp<ViewStyle>;
  bodyStyle?: StyleProp<ViewStyle>;
  footerStyle?: StyleProp<ViewStyle>;
  disabled?: boolean;
  onCancel?: () => void;
  onConfirm?: () => void;
  title?: string;
  titleStyle?: StyleProp<TextStyle>;
  message?: string;
  messageStyle?: StyleProp<TextStyle>;
  delay?: number;
  initialVisible?: boolean;
};

export type ModalDialogHandle = {
  show: () => void;
  hide: () => void;
  toggle: () => void;
};

const {height: WINDOW_HEIGHT} = Dimensions.get('window');

export const defaultProps = {
  backgroundColor: 'white',
  animationType: Platform.select<ModalProps['animationType']>({android: 'fade', ios: 'slide'})!,
  cancelTitle: 'Cancel',
  confirmTitle: Platform.select<string>({android: 'OK', ios: 'Confirm'})!,
  transparent: true
};

export const ModalDialog = forwardRef<ModalDialogHandle, ModalDialogProps>(
  (
    {
      animationType = defaultProps.animationType,
      backgroundColor = defaultProps.backgroundColor,
      cancelTitle = defaultProps.cancelTitle,
      cancelStyle,
      confirmTitle = defaultProps.confirmTitle,
      confirmStyle,
      bodyStyle,
      modalStyle,
      containerStyle,
      children,
      delay = 0,
      disabled = false,
      onCancel,
      onConfirm,
      title,
      titleStyle,
      message,
      messageStyle,
      headerStyle,
      footerStyle,
      transparent = defaultProps.transparent,
      initialVisible = false,
      defaultStyles: propDefaultStyles = defaultStyles,
      ...otherModalProps
    },
    ref
  ) => {
    const backdrop = useContext<BackdropContextValue>(BackdropContext);
    const [isVisible, setIsVisible] = useState(initialVisible);
    const latestIsVisible = useRef<boolean>(isVisible);

    // Show modal
    const show = useCallback(() => {
      if (disabled) {
        return;
      }
      setIsVisible(true);
    }, [disabled]);

    // Hide modal
    const hide = useCallback(() => {
      if (disabled) {
        return;
      }
      setIsVisible(false);
    }, [disabled]);

    // Track isVisible latest value with a ref
    useEffect(() => {
      latestIsVisible.current = isVisible;
    }, [isVisible]);

    // Toggle modal
    // @NOTE a ref is used to provide a stable context accross renders
    const toggle = useCallback(() => {
      !latestIsVisible.current ? show() : hide();
    }, [latestIsVisible, hide, show]);

    // Expose API via an imperative handle
    useImperativeHandle(ref, () => ({show, hide, toggle}), [show, hide, toggle]);

    // Animate backdrop along "isVisible" prop
    useEffect(() => {
      if (backdrop) {
        isVisible ? backdrop.show() : backdrop.hide();
      }
    }, [backdrop, isVisible]);

    // Toggle keyboard along "isVisible" prop
    useEffect(() => {
      // Force blur on other inputs
      if (isVisible) {
        Keyboard.dismiss();
      }
    }, [isVisible]);

    // Always hide on unmount
    useEffect(() => {
      return () => {
        hide();
      };
    }, [hide]);

    // Close the modal and propagate cancel action
    const handleModalCancel = useCallback(() => {
      if (onCancel) {
        onCancel();
      }
      hide();
    }, [onCancel, hide]);

    // Close the modal and propagate confirm action
    const handleModalConfirm = useCallback(() => {
      if (onConfirm) {
        setTimeout(() => {
          onConfirm();
        }, delay);
      }
      hide();
    }, [onConfirm, hide, delay]);

    const isHeaderVisible = title || message;
    const isFooterVisible = (Platform.OS === 'android' && onCancel) || onConfirm;
    const inheritedBodyStyle = useMemo<ViewStyle>(() => {
      const inheritedStyles = {};
      if (!isHeaderVisible) {
        Object.assign(inheritedStyles, {
          borderTopLeftRadius: propDefaultStyles.header.borderTopLeftRadius,
          borderTopRightRadius: propDefaultStyles.header.borderTopRightRadius
        });
      }
      if (!isFooterVisible) {
        Object.assign(inheritedStyles, {
          borderBottomLeftRadius: propDefaultStyles.footer.borderBottomLeftRadius,
          borderBottomRightRadius: propDefaultStyles.footer.borderBottomRightRadius
        });
      }
      return inheritedStyles;
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isHeaderVisible, isFooterVisible]);

    return (
      <Modal visible={isVisible} animationType={animationType} transparent={transparent} {...otherModalProps}>
        <TouchableWithoutFeedback style={{flex: 1}} onPress={handleModalCancel}>
          <View style={[propDefaultStyles.modal, modalStyle]}>
            <TouchableWithoutFeedback style={{flex: 1}}>
              <SafeAreaView style={[propDefaultStyles.container, containerStyle]}>
                {isHeaderVisible ? (
                  <View style={[propDefaultStyles.header, {backgroundColor}, headerStyle]}>
                    {title ? <Text style={[propDefaultStyles.title, titleStyle]}>{title}</Text> : null}
                    {message ? <Text style={[propDefaultStyles.message, messageStyle]}>{message}</Text> : null}
                  </View>
                ) : null}
                <View style={[propDefaultStyles.body, inheritedBodyStyle, {backgroundColor}, bodyStyle]}>
                  {children}
                </View>
                {isFooterVisible ? (
                  <View style={[propDefaultStyles.footer, {backgroundColor}, footerStyle]}>
                    {Platform.OS === 'android' && onCancel ? (
                      <ModalDialogButton
                        style={[propDefaultStyles.cancel, cancelStyle]}
                        onPress={handleModalCancel}
                        title={cancelTitle}
                      />
                    ) : null}
                    {onConfirm ? (
                      <ModalDialogButton
                        style={[propDefaultStyles.confirm, confirmStyle]}
                        onPress={handleModalConfirm}
                        title={confirmTitle}
                      />
                    ) : null}
                  </View>
                ) : null}
                {Platform.OS === 'ios' && onCancel ? (
                  <ModalDialogButton
                    style={[propDefaultStyles.cancel, {backgroundColor}, cancelStyle]}
                    onPress={handleModalCancel}
                    title={cancelTitle}
                  />
                ) : null}
              </SafeAreaView>
            </TouchableWithoutFeedback>
          </View>
        </TouchableWithoutFeedback>
      </Modal>
    );
  }
);

export const defaultStyles = {
  modal: Platform.select<ViewStyle>({
    ios: {flex: 1, justifyContent: 'flex-end'},
    android: {flex: 1, justifyContent: 'center'}
  })!,
  container: Platform.select<ViewStyle>({
    ios: {
      flexDirection: 'column',
      alignItems: 'stretch',
      marginHorizontal: 8,
      marginBottom: 8,
      maxHeight: WINDOW_HEIGHT - 8
    },
    android: {
      flexDirection: 'column',
      alignItems: 'stretch',
      // marginHorizontal: 82, // timepicker/spinner
      marginHorizontal: 24,
      elevation: 4,
      maxHeight: WINDOW_HEIGHT - 8
    }
  })!,
  header: Platform.select<ViewStyle>({
    ios: {alignItems: 'center', padding: 16, borderTopLeftRadius: 12, borderTopRightRadius: 12},
    android: {alignItems: 'flex-start', padding: 16}
  })!,
  body: Platform.select<ViewStyle>({
    ios: {
      flexShrink: 1,
      flexDirection: 'column'
    },
    android: {
      flexShrink: 1,
      flexDirection: 'column'
    }
  })!,
  footer: Platform.select<TextStyle>({
    ios: {
      borderBottomLeftRadius: 12,
      borderBottomRightRadius: 12
    },
    android: {
      flexDirection: 'row',
      justifyContent: 'flex-end',
      borderBottomLeftRadius: 2,
      borderBottomRightRadius: 2,
      padding: 8
    }
  })!,
  title: Platform.select<TextStyle>({
    ios: {paddingBottom: 12, fontSize: 14, fontWeight: '500', textAlign: 'center', color: '#888'},
    android: {
      paddingVertical: 10,
      paddingHorizontal: 12,
      fontSize: 20,
      fontWeight: '500',
      textAlign: 'left',
      color: '#333'
    }
  })!,
  message: Platform.select<TextStyle>({
    ios: {paddingBottom: 12, fontSize: 13, fontWeight: '400', textAlign: 'center', color: '#888'},
    android: {paddingVertical: 6, paddingHorizontal: 12, fontSize: 18, fontWeight: '400', color: '#666'}
  })!,
  cancel: Platform.select<TextStyle>({
    ios: {marginTop: 24, borderRadius: 12, fontWeight: '600'},
    android: {}
  })!,
  confirm: Platform.select<TextStyle>({
    ios: {borderBottomLeftRadius: 12, borderBottomRightRadius: 12, fontWeight: '400'},
    android: {}
  })!
};
