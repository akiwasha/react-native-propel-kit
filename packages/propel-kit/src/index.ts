export * from '@mgcrea/react-native-action-sheet-provider';
export * from '@mgcrea/react-native-backdrop-provider';
export * from '@mgcrea/react-native-button';
export * from '@mgcrea/react-native-date-picker';
export * from '@mgcrea/react-native-modal-dialog';
export * from '@mgcrea/react-native-select';
