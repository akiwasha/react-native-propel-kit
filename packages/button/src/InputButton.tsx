import React, {FunctionComponent} from 'react';
import {Platform, StyleProp, TextInputProps, TextStyle, ViewStyle} from 'react-native';
import {Pressable, PressableProps} from './Pressable';

export type InputButtonProps = Omit<TextInputProps, 'onPress'> & {
  onFocus: PressableProps['onPress'];
  viewStyle?: StyleProp<ViewStyle>;
  disabledViewStyle?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  disabledTextStyle?: StyleProp<TextStyle>;
};

export const defaultProps = {
  placeholderTextColor: Platform.select({ios: '#c7c7cc', android: '#9e9e9e'})
};

export const InputButton: FunctionComponent<InputButtonProps> = ({
  editable = true,
  onFocus,
  placeholder,
  viewStyle,
  disabledViewStyle,
  textStyle,
  disabledTextStyle,
  placeholderTextColor = defaultProps.placeholderTextColor,
  value,
  ...otherPressableProps
}) => {
  return (
    <Pressable
      onPress={onFocus}
      title={value || placeholder || ''}
      disabled={!editable}
      viewStyle={[defaultStyles.view, viewStyle]}
      disabledViewStyle={[defaultStyles.disabledView, disabledViewStyle]}
      textStyle={[defaultStyles.text, textStyle]}
      disabledTextStyle={[defaultStyles.disabledText, disabledTextStyle]}
      color={!value ? placeholderTextColor : undefined}
      numberOfLines={1}
      {...otherPressableProps}
    />
  );
};

export const defaultStyles: {[s: string]: ViewStyle | TextStyle} = {
  view: Platform.select<ViewStyle>({
    ios: {justifyContent: 'flex-start', alignItems: 'center'},
    android: {justifyContent: 'flex-start', height: 50}
  })!,
  disabledView: Platform.select<ViewStyle>({
    ios: {},
    android: {}
  })!,
  text: Platform.select<TextStyle>({
    ios: {fontSize: 14, color: 'black'},
    android: {alignSelf: 'flex-start', paddingHorizontal: 3, paddingBottom: 2, color: '#212121'}
  })!,
  disabledText: Platform.select<TextStyle>({
    ios: {color: '#cdcdcd'},
    android: {color: '#cdcdcd'}
  })!
};
