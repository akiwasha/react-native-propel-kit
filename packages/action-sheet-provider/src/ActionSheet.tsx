import ModalDialog, {ModalDialogHandle, ModalDialogProps} from '@mgcrea/react-native-modal-dialog';
import React, {forwardRef, ForwardRefRenderFunction, useMemo} from 'react';
import {
  ActionSheetIOSOptions,
  Dimensions,
  Platform,
  ScrollView,
  ScrollViewProps,
  StyleProp,
  StyleSheet,
  TextStyle,
  ViewStyle
} from 'react-native';
import ActionSheetOption from './components/ActionSheetOption';
import isUndefined from './utils/isUndefined';

const {height: WINDOW_HEIGHT} = Dimensions.get('window');
const IOS_OPACITY = 0.85; // @NOTE from native ActionSheet (though it uses blur)

export type ActionSheetProps = Pick<ScrollViewProps, 'scrollEnabled'> &
  Omit<ModalDialogProps, 'onConfirm' | 'confirmTitle' | 'confirmStyle'> &
  ActionSheetIOSOptions & {
    defaultStyles?: typeof defaultStyles;
    optionStyle?: StyleProp<TextStyle>;
    lastOptionExtraStyle?: StyleProp<TextStyle>;
    destructiveButtonColor?: string;
    onButtonPress: (buttonIndex: number) => void;
    onCancel: () => void;
  };

export const defaultProps = {
  backgroundColor: 'white',
  cancelTitle: 'Cancel',
  destructiveButtonColor: Platform.select({
    ios: '#ff3b30', // iOS.systemRed (@see https://developer.apple.com/design/human-interface-guidelines/ios/visual-design/color/)
    android: '#f44336' // android.red500 (@see https://material.io/design/color/the-color-system.html)
  })!
};

const ActionSheetRenderFunction: ForwardRefRenderFunction<ModalDialogHandle, ActionSheetProps> = (
  {
    options,
    backgroundColor = defaultProps.backgroundColor,
    optionStyle: propOptionStyle,
    lastOptionExtraStyle: propLastOptionExtraStyle,
    cancelButtonIndex,
    destructiveButtonIndex,
    destructiveButtonColor = defaultProps.destructiveButtonColor,
    onButtonPress,
    scrollEnabled: propScrollEnabled,
    defaultStyles: propDefaultStyles = defaultStyles,
    ...otherModalDialogProps
  },
  ref
) => {
  const cancelTitle = !isUndefined(cancelButtonIndex) && options[cancelButtonIndex] ? options[cancelButtonIndex] : '';
  const scrollEnabled = !isUndefined(cancelButtonIndex) ? options.length > 6 : propScrollEnabled;

  const optionStyle = useMemo<ViewStyle>(
    () => StyleSheet.flatten([propDefaultStyles.option, {backgroundColor}, propOptionStyle]),
    [propDefaultStyles, backgroundColor, propOptionStyle]
  );
  const lastOptionExtraStyle = useMemo<ViewStyle>(
    () => StyleSheet.flatten([propDefaultStyles.lastOptionExtra, propLastOptionExtraStyle]),
    [propDefaultStyles, propLastOptionExtraStyle]
  );

  return (
    <ModalDialog
      ref={ref}
      defaultStyles={propDefaultStyles}
      backgroundColor={backgroundColor}
      cancelTitle={cancelTitle}
      bodyStyle={{backgroundColor: 'transparent'}}
      {...otherModalDialogProps}>
      <ScrollView scrollEnabled={scrollEnabled}>
        {options.map((buttonTitle, buttonIndex) => {
          // Skip cancel button
          if (!isUndefined(cancelButtonIndex) && buttonIndex === cancelButtonIndex) {
            return null;
          }

          const color: string =
            destructiveButtonIndex && buttonIndex === destructiveButtonIndex ? destructiveButtonColor : '';
          return (
            <ActionSheetOption
              style={[optionStyle as TextStyle].concat(
                color ? [{color}] : [],
                buttonIndex === options.length - 1 ? [lastOptionExtraStyle] : []
              )}
              onPress={() => onButtonPress(buttonIndex)}
              title={buttonTitle}
              key={`${buttonIndex}`} // eslint-disable-line react/no-array-index-key
            />
          );
        })}
      </ScrollView>
    </ModalDialog>
  );
};

export const ActionSheet = forwardRef(ActionSheetRenderFunction);

export const defaultStyles = {
  modal: Platform.select<ViewStyle>({
    ios: {flex: 1, justifyContent: 'flex-end'},
    android: {flex: 1, justifyContent: 'center'}
  })!,
  container: Platform.select<ViewStyle>({
    ios: {
      flexDirection: 'column',
      alignItems: 'stretch',
      marginHorizontal: 8,
      marginBottom: 8,
      maxHeight: WINDOW_HEIGHT - 8
    },
    android: {
      flexDirection: 'column',
      alignItems: 'stretch',
      // marginHorizontal: 82, // timepicker/spinner
      margin: 24,
      maxHeight: WINDOW_HEIGHT - 24 * 2,
      elevation: 4
    }
  })!,
  header: Platform.select<ViewStyle>({
    ios: {
      alignItems: 'center',
      padding: 16,
      borderTopLeftRadius: 12,
      borderTopRightRadius: 12,
      opacity: IOS_OPACITY
    },
    android: {alignItems: 'flex-start', paddingVertical: 24, paddingHorizontal: 32}
  })!,
  body: Platform.select<ViewStyle>({
    ios: {
      flexShrink: 1,
      flexDirection: 'column',
      backgroundColor: 'transparent',
      opacity: IOS_OPACITY,
      borderBottomStartRadius: 12,
      borderBottomEndRadius: 12
    },
    android: {
      flexShrink: 1,
      flexDirection: 'column',
      backgroundColor: 'transparent'
    }
  })!,
  footer: Platform.select<TextStyle>({
    ios: {}, // @NOTE we don't have a footer on iOS
    android: {
      flexDirection: 'row',
      justifyContent: 'flex-end',
      borderBottomLeftRadius: 2,
      borderBottomRightRadius: 2,
      padding: 8
    }
  })!,
  title: Platform.select<TextStyle>({
    ios: {paddingBottom: 12, fontSize: 14, fontWeight: '500', textAlign: 'center', color: '#888'},
    android: {
      paddingBottom: 12,
      fontSize: 20,
      fontWeight: '500',
      textAlign: 'left',
      color: '#333'
    }
  })!,
  message: Platform.select<TextStyle>({
    ios: {paddingBottom: 12, fontSize: 13, fontWeight: '400', textAlign: 'center', color: '#888'},
    android: {paddingBottom: 12, fontSize: 18, fontWeight: '400', color: '#666'}
  })!,
  cancel: Platform.select<TextStyle>({
    ios: {marginTop: 8, borderRadius: 12, fontWeight: '600'},
    android: {}
  })!,
  confirm: Platform.select<TextStyle>({
    ios: {}, // @NOTE we don't have a confirm on iOS
    android: {}
  })!,
  option: Platform.select<TextStyle>({
    ios: {
      marginTop: StyleSheet.hairlineWidth
    },
    android: {
      marginTop: StyleSheet.hairlineWidth
    }
  })!,
  lastOptionExtra: Platform.select<TextStyle>({
    ios: {
      borderBottomLeftRadius: 12,
      borderBottomRightRadius: 12
    },
    android: {
      marginBottom: StyleSheet.hairlineWidth
    }
  })!
};
