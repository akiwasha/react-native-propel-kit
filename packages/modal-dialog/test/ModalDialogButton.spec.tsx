import React from 'react';
import 'react-native';
import {render, fireEvent} from 'react-native-testing-library';

import {ModalDialogButton} from './../src';

describe('<ModalDialogButton>', () => {
  it('renders correctly', () => {
    const component = render(<ModalDialogButton title="My Title" onPress={() => {}} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe('onPress prop', () => {
    it('should properly call callback', () => {
      const onPressSpy = jest.fn();
      const {getByType, toJSON} = render(<ModalDialogButton title="My Title" onPress={onPressSpy} />);
      expect(toJSON()).toMatchSnapshot();
      fireEvent.press(getByType(ModalDialogButton));
      expect(onPressSpy).toHaveBeenCalledTimes(1);
    });
  });
});
