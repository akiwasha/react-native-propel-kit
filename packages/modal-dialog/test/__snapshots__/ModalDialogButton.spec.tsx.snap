// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalDialogButton> onPress prop should properly call callback 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 0,
      "flexDirection": "row",
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexGrow": 1,
        "height": 56,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#007aff",
          "fontSize": 20,
        }
      }
    >
      My Title
    </Text>
  </View>
</View>
`;

exports[`<ModalDialogButton> renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 0,
      "flexDirection": "row",
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexGrow": 1,
        "height": 56,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#007aff",
          "fontSize": 20,
        }
      }
    >
      My Title
    </Text>
  </View>
</View>
`;
